File: sdks/javascript/src/models/file-upload.js
/**
 * File upload models for the Hive Vectorizer SDK.
 */

/**
 * @typedef {Object} FileUploadRequest
 * @property {string} collectionName - Target collection name
 * @property {number} [chunkSize] - Chunk size in characters
 * @property {number} [chunkOverlap] - Chunk overlap in characters
 * @property {Object} [metadata] - Additional metadata to attach to all chunks
 */

/**
 * @typedef {Object} FileUploadResponse
 * @property {boolean} success - Whether the upload was successful
 * @property {string} filename - Original filename
 * @property {string} collection_name - Target collection
 * @property {number} chunks_created - Number of chunks created from the file
 * @property {number} vectors_created - Number of vectors created and stored
 * @property {number} file_size - File size in bytes
 * @property {string} language - Detected language/file type
 * @property {number} processing_time_ms - Processing time in milliseconds
 */

/**
 * @typedef {Object} FileUploadConfig
 * @property {number} max_file_size - Maximum file size in bytes
 * @property {number} max_file_size_mb - Maximum file size in megabytes
 * @property {string[]} allowed_extensions - List of allowed file extensions
 * @property {boolean} reject_binary - Whether binary files are rejected
 * @property {number} default_chunk_size - Default chunk size in characters
 * @property {number} default_chunk_overlap - Default chunk overlap in characters
 */

/**
 * @typedef {Object} UploadFileOptions
 * @property {number} [chunkSize] - Chunk size in characters
 * @property {number} [chunkOverlap] - Chunk overlap in characters
 * @property {Object} [metadata] - Additional metadata to attach to all chunks
 */

/**
 * Validate a FileUploadResponse object.
 * @param {Object} response - Response to validate
 * @throws {Error} If validation fails
 */
export function validateFileUploadResponse(response) {
  if (!response || typeof response !== 'object') {
    throw new Error('Invalid file upload response');
  }
  if (typeof response.success !== 'boolean') {
    throw new Error('File upload response must have a boolean success field');
  }
  if (typeof response.filename !== 'string') {
    throw new Error('File upload response must have a string filename field');
  }
  if (typeof response.collection_name !== 'string') {
    throw new Error('File upload response must have a string collection_name field');
  }
}

/**
 * Validate a FileUploadConfig object.
 * @param {Object} config - Config to validate
 * @throws {Error} If validation fails
 */
export function validateFileUploadConfig(config) {
  if (!config || typeof config !== 'object') {
    throw new Error('Invalid file upload config');
  }
  if (typeof config.max_file_size !== 'number') {
    throw new Error('File upload config must have a number max_file_size field');
  }
  if (!Array.isArray(config.allowed_extensions)) {
    throw new Error('File upload config must have an array allowed_extensions field');
  }
}

File: sdks/rust/src/models/file_upload.rs
//! File upload models for the Vectorizer SDK

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

/// Request to upload a file for indexing
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileUploadRequest {
    /// Target collection name
    pub collection_name: String,
    /// Chunk size in characters (uses server default if not specified)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub chunk_size: Option<u32>,
    /// Chunk overlap in characters (uses server default if not specified)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub chunk_overlap: Option<u32>,
    /// Additional metadata to attach to all chunks
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

/// Response from file upload operation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileUploadResponse {
    /// Whether the upload was successful
    pub success: bool,
    /// Original filename
    pub filename: String,
    /// Target collection
    pub collection_name: String,
    /// Number of chunks created from the file
    pub chunks_created: u32,
    /// Number of vectors created and stored
    pub vectors_created: u32,
    /// File size in bytes
    pub file_size: u64,
    /// Detected language/file type
    pub language: String,
    /// Processing time in milliseconds
    pub processing_time_ms: u64,
}

/// Configuration for file uploads
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileUploadConfig {
    /// Maximum file size in bytes
    pub max_file_size: u64,
    /// Maximum file size in megabytes
    pub max_file_size_mb: u32,
    /// List of allowed file extensions
    pub allowed_extensions: Vec<String>,
    /// Whether binary files are rejected
    pub reject_binary: bool,
    /// Default chunk size in characters
    pub default_chunk_size: u32,
    /// Default chunk overlap in characters
    pub default_chunk_overlap: u32,
}

/// Options for uploading a file
#[derive(Debug, Clone, Default)]
pub struct UploadFileOptions {
    /// Chunk size in characters
    pub chunk_size: Option<u32>,
    /// Chunk overlap in characters
    pub chunk_overlap: Option<u32>,
    /// Additional metadata to attach to all chunks
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}
